core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosState.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.mesos
+
+import org.apache.mesos.state._
+import scala.concurrent.duration.Duration
+import java.util.concurrent._
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.util.Utils
+import java.lang.Boolean
+import java.util
+import java.util.Collections
+import org.apache.mesos.MesosNativeLibrary
+import org.apache.spark.deploy.SparkCuratorUtil
+
+/**
+ * An abstraction over Mesos state abstraction. This to provide automatic retries and
+ * serialization of objects, and translating state exceptions into Spark exceptions.
+ */
+private[spark] class MesosState(conf: SparkConf) {
+  val defaultFetchTimeoutMs = conf.getLong("spark.mesos.state.fetch.timeout.ms", 2000)
+  val defaultStoreTimeoutMs = conf.getLong("spark.mesos.state.store.timeout.ms", 5000)
+  val retries = conf.getInt("spark.mesos.state.store.retries", 3)
+  val quorum = conf.getInt("spark.mesos.state.quorum", 1)
+  val path = conf.get("spark.mesos.state.path", "/.spark_mesos_dispatcher")
+  private val ZK_SESSION_TIMEOUT_MILLIS = 60000
+
+  private val state = conf.get("spark.deploy.recoveryMode", "NONE").toUpperCase() match {
+    case "NONE" => new InMemoryState()
+    case "ZOOKEEPER" => {
+      val servers = conf.get("spark.deploy.zookeeper.url")
+      new ZooKeeperState(
+        servers,
+        ZK_SESSION_TIMEOUT_MILLIS,
+        TimeUnit.MILLISECONDS,
+        "/spark_mesos_dispatcher")
+    }
```

Comment:
Same here, this match is not exhaustive. You will need to add a `case _ =>` that either defaults to `InMemoryState` or throws an informative exception.